src/main/java/org/candlepin/subscriptions/capacity/CapacityReconciliationController.java

```diff
@@ -0,0 +1,118 @@
+  private final CapacityProductExtractor productExtractor;
+  private final SubscriptionCapacityRepository subscriptionCapacityRepository;
+  private final Counter capacityRecordsCreated;
+  private final Counter capacityRecordsUpdated;
+  private final Counter capacityRecordsDeleted;
+
+  public CapacityReconciliationController(
+      OfferingRepository offeringRepository,
+      ProductWhitelist productWhitelist,
+      CapacityProductExtractor productExtractor,
+      SubscriptionCapacityRepository subscriptionCapacityRepository,
+      MeterRegistry meterRegistry) {
+    this.offeringRepository = offeringRepository;
+    this.productWhitelist = productWhitelist;
+    this.productExtractor = productExtractor;
+    this.subscriptionCapacityRepository = subscriptionCapacityRepository;
+    capacityRecordsCreated = meterRegistry.counter("rhsm-subscriptions.capacity.records_created");
+    capacityRecordsUpdated = meterRegistry.counter("rhsm-subscriptions.capacity.records_updated");
+    capacityRecordsDeleted = meterRegistry.counter("rhsm-subscriptions.capacity.records_deleted");
+  }
+
+  public void reconcileCapacityForSubscription(Subscription subscription) {
+
+    Collection<SubscriptionCapacity> newCapacities = mapSubscriptionToCapacities(subscription);
+    reconcileSubscriptionCapacities(
+        newCapacities, subscription.getSubscriptionId(), subscription.getSku());
+  }
+
+  private Collection<SubscriptionCapacity> mapSubscriptionToCapacities(Subscription subscription) {
+
+    Offering offering = offeringRepository.getById(subscription.getSku());
+    Set<String> products =
+        productExtractor.getProducts(
+            offering.getProductIds().stream().map(Object::toString).collect(Collectors.toSet()));
+
+    return products.stream()
+        .map(product -> from(subscription, offering, product))
+        .collect(Collectors.toList());
+  }
+
+  private void reconcileSubscriptionCapacities(
+      Collection<SubscriptionCapacity> newCapacities, String subscriptionId, String sku) {
+
+    Collection<SubscriptionCapacity> toSave = new ArrayList<>();
+    Map<SubscriptionCapacityKey, SubscriptionCapacity> existingCapacityMap =
+        subscriptionCapacityRepository.findByKeySubscriptionId(subscriptionId).stream()
+            .collect(Collectors.toMap(SubscriptionCapacity::getKey, Function.identity()));
+
+    if (productWhitelist.productIdMatches(sku)) {
+      newCapacities.forEach(
+          newCapacity -> {
+            toSave.add(newCapacity);
+            SubscriptionCapacity oldVersion = existingCapacityMap.remove(newCapacity.getKey());
+            if (oldVersion != null) {
+              capacityRecordsUpdated.increment();
+            } else {
+              capacityRecordsCreated.increment();
+            }
+          });
+      subscriptionCapacityRepository.saveAll(toSave);
+    }
+
+    Collection<SubscriptionCapacity> toDelete = new ArrayList<>(existingCapacityMap.values());
```

Comment:
Ah, ok so I see what was up with the map that you were removing elements from....my comment about groupingBy doesn't take that into consideration at all

Reply:
@lindseyburnett Are there any changes that you recommend that I make here? I

Reply:
@snehagunta  nope, I'm good...it was all related to projections and stuff what we're tabling